Dev guide contextualization | Content release #1

flossypurse opened the pull request:
**New content**
- About Temporal SDKs: https://temporal-documentation-git-edu-1036-release-01.preview.thundergun.io/dev-guide/sdks

**New content + reorg**
- Introduction to the Go SDK: https://temporal-documentation-git-edu-1036-release-01.preview.thundergun.io/dev-guide/go/introduction
- Introduction to the Java SDK: https://temporal-documentation-git-edu-1036-release-01.preview.thundergun.io/dev-guide/java/introduction
- Introduction to the Python SDK: https://temporal-documentation-git-edu-1036-release-01.preview.thundergun.io/dev-guide/python/introduction
- Introduction to the TypeScript SDK: https://temporal-documentation-git-edu-1036-release-01.preview.thundergun.io/dev-guide/typescript/introduction

angelazhou32 commented on assembly/guide-configs/app-dev/sdks.json:
```diff
@@ -0,0 +1,39 @@
+    "title": "About Temporal SDKs",
+    "sidebar_label": "About the SDKs",
+    "sidebar_position": 1,
+    "description": "Explore the components that make up a Temporal SDK and how it works to create a durable execution.",
```
*how they work to create a durable execution?

angelazhou32 commented on docs-src/dev-guide/major-components.md:
````diff
@@ -0,0 +1,263 @@
+}
+```
+
+A Workflow executes Activities (other functions), handles and sends messages (Queries, Signals, Updates), and interacts with other Workflows.
````
"other functions" is a bit vague to me..how about other functions that interact with external systems, functions that work with non-deterministic code etc.

angelazhou32 commented on docs-src/dev-guide/why-use-a-temporal-sdk.md:
```diff
@@ -0,0 +1,74 @@
+
+:::caution Do not copy and use code
+
+The following is sudo code, for tested samples see your langauge SDK's developer's guide.
```
psuedocode

angelazhou32 commented on docs-src/dev-guide/why-use-a-temporal-sdk.md:
```diff
@@ -0,0 +1,74 @@
+
+:::caution Do not copy and use code
+
+The following is sudo code, for tested samples see your langauge SDK's developer's guide.
```
psuedocode

dandavison commented on docs-src/dev-guide/major-components.md:
```diff
@@ -0,0 +1,263 @@
+
+**What are the major components of Temporal SDKs?**
+
+Temporal SDKs offer developers with the following:
```
Temporal SDKs offer developers ~~with~~ the following:

dandavison commented on docs-src/dev-guide/major-components.md:
```diff
@@ -0,0 +1,263 @@
+- A Temporal Client to communicate with a Temporal Cluster
+- APIs to develop application code Workflows & Activities
+- APIs to configure and run Workers
+- A common library for code that’s used across the Client, Worker, and Workflow
```
This is slightly confusing because it reads like it's describing something distinct from the other bulletpoints, but I don't think it actually is.

dandavison commented on docs-src/dev-guide/major-components.md:
```diff
@@ -0,0 +1,263 @@
+
+### Temporal Client
+
+A Temporal Client acts as a bridge of communication between the systems outside of Temporal and your application.
```
"outside of Temporal" seems wrong. How about

> A Temporal Client acts as the bridge for communication between your application and the Temporal server.

dandavison commented on docs-src/dev-guide/major-components.md:
```diff
@@ -0,0 +1,263 @@
+
+### APIs to Develop Workflows
+
+Workflows are defined as code as either a function or an object method, depending on the language.
```
How about

> Workflows are defined as code: either a function or an object method, depending on the language.

tomwheeler commented on docs-src/dev-guide/major-components.md:
```diff
@@ -0,0 +1,263 @@
+	applicantDetails := ApplicantDetails{
+		// ...
+	}
+  // Start the Workflow
```
This comment is improperly indented, as are the comments above it (lines 50 and 56)

tomwheeler commented on docs-src/dev-guide/major-components.md:
```diff
@@ -0,0 +1,263 @@
+
+Developers can then use the Client as the main entry point for interacting with the application through Temporal.
+From here, developers can now start or Signal Workflows, query the Workflow state etc.
+We can see in the example above how the developer has used the provided start method to start the Workflow upon the running of the function.
```
> how the developer has used the provided start method

It was unclear to me what you mean by "provided start method." Are you referring to the entire program (which I assume would be named `start/main.go`), the `client.ExecuteWorkflow` call, or something else? Regardless, I would avoid the word "method" since that may be interpreted in the object oriented way. Maybe "has used provided means of starting the Workflow" would be more clear.

mjameswh replied:
> to start ~the~ Workflow ~upon the running of the function~.

to start **a** Workflow.

tomwheeler commented on docs-src/go/introduction-to-go-sdk.md:
```diff
@@ -0,0 +1,129 @@
+
+:::
+
+Temporal Python SDK first released on September 29, 2020, and has been enabling Python developers to build Temporal Applications ever since.
```
> Temporal Python SDK first released on September 29, 2020, and has been enabling Python developers to build Temporal Applications ever since.

I call shenanigans on this claim. 

According to the [releases page in GitHub](https://github.com/temporalio/sdk-python/releases), the Python SDK was first released (as alpha 0.1a1) on March 18, 2022. In fact, [the first commit](https://github.com/temporalio/sdk-python/commit/75f7113428f64c8a2c3f3f540f005331a2d0fa37) in the repo was only about two months earlier, or well over a year after the September 29, 2020 initial release date claimed here.

tomwheeler replied:
Of course, the bigger issue is that this is talking about the release of the Python SDK in a document intended for Go developers. Although I would argue that initial release date of the Python SDK isn't even relevant to a Python developer, it's certainly not relevant to a Go developer.

tomwheeler commented on docs-src/go/introduction-to-go-sdk.md:
```diff
@@ -0,0 +1,129 @@
+
+Further resources for learning how to use the Go SDK include the following:
+
+- [Temporal 101 with Go](https://learn.temporal.io/courses/temporal_101/go): our free introductory course.
```
I would also recommend Temporal 102 with Go.

tomwheeler commented on docs/dev-guide/golang/introduction.md:
```diff
@@ -0,0 +1,136 @@
+
+:::
+
+Temporal Python SDK first released on September 29, 2020, and has been enabling Python developers to build Temporal Applications ever since.
```
This info about the Python SDK release date is not relevant, inaccurate, and seems to be copied and pasted in multiple locations.

dandavison commented on docs-src/dev-guide/sdk-and-cluster-relationship.md:
```diff
@@ -0,0 +1,101 @@
+
+![The Worker picks up the new Task](/diagrams/how-sdk-works-1.svg)
+
+This is when the SDK Worker Replays the Workflow code, uses the Event History as guidance on what to expect. If the Replay encounters an Event that doesn’t match up with what is expected from the code, a non determinism error is thrown.
```
> Source: [definitive documentation on errors](https://docs.temporal.io/references/errors#non-deterministic-error)

That's unfortunate -- "non-determinism error" is better than "non-deterministic error"! The latter sounds like an adjective indicating that the error itself is non-deterministic, whereas the former makes it clearer that the error is indicating that some non-determinism has been detected.

flossypurse replied:
added hyphen, kept "determinism" because it is meant to be more generic in this setting.

mjameswh commented on docs-src/dev-guide/major-components.md:
````diff
@@ -0,0 +1,263 @@
+```
+
+Developers can then use the Client as the main entry point for interacting with the application through Temporal.
+From here, developers can now start or Signal Workflows, query the Workflow state etc.
````
```suggestion
Using that Client, developers may for example start or signal Workflows, query a Workflow state, etc.
```

flossypurse replied:
👍

flossypurse commented on docs-src/dev-guide/major-components.md:
```diff
@@ -0,0 +1,263 @@
+Temporal SDKs offer developers with the following:
+
+- A Temporal Client to communicate with a Temporal Cluster
+- APIs to develop application code Workflows & Activities
```
Workflows and Activities should probably have been in paranthesis here.
- APIs to develop application code (Workflows and Activities).

flossypurse commented on docs-src/dev-guide/sdk-and-cluster-relationship.md:
```diff
@@ -0,0 +1,101 @@
+
+The Temporal Cluster functions more as a choreographer than a conductor. Rather than directly assigning tasks to Workers, the Cluster arranges the Tasks into a Task Queue while Workers poll the Task Queue. Developers may create a fleet of Workers and tune them so that a Task is picked up as soon as it is available. If a Worker goes down, Tasks can wait until the next Worker is available.
+
+A Workflow might request to execute an Activity, start a Timer, or start a Child Workflow, each of which translates into Commands, dispatched to the Temporal Cluster, at which point the Cluster translates the Commands into Events that become part of a Workflow Execution’s Event History.
```
Dig it.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/temporalio/documentation?pullRequest=2327) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/temporalio/documentation?pullRequest=2327) before we can accept your contribution.<br/>**4** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: rachfop<br/>:white_check_mark: djmagee<br/>:white_check_mark: flossypurse<br/>:white_check_mark: DJSanti<br/>:x: axfelix<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/temporalio/documentation?pullRequest=2327) it.</sub>

Sushisource commented on docs-src/java/introduction-to-java-sdk.md:
```diff
@@ -0,0 +1,162 @@
+- Interfaces
+- Inheritance
+- Encapsulation
+- Polymorphism
```
+1 - I would prefer us to have a "recommended material" section for general programming advice or similar, and just link people out to some books / sites / etc that make this their core business.

flossypurse replied:
Totally acknowledged.
We want to try out being much more prescriptive all around and see how this goes.

flossypurse commented on docs-src/dev-guide/temporal-application.md:
```diff
@@ -0,0 +1,25 @@
+- Recoverable: The ability of a process to resume execution after suspending due to a _failure_.
+- Reactive: The ability of a process to respond to external events.
+
+Hence, a Temporal Workflow Execution executes a [Temporal Workflow Definition](/concepts/what-is-a-workflow-definition), also known as a Temporal Workflow Function or your application code, exactly once and to completion, regardless of whether your code runs for seconds or years and in the presence of arbitrary load and failures.
```
Re-writing this section

flossypurse commented on docs-src/dev-guide/temporal-sdk.md:
```diff
@@ -0,0 +1,16 @@
+
+SDKs are more than just a development tool, however.
+The SDK APIs enable developers to write code in a particular pattern that mirrors real world processes.
+And the SDK's internal implementation, working in collaboration with the Temporal Cluster, steps through that code, guaranteeing execution progression during application runtime.
```
Acknowledged, though I am failing to meet your perspective on this one.

dandavison commented:
I've only left negative comments so far :) Tons of great content here!

tomwheeler reviewed (approved):
My suggestions with the prose were addressed and I am satisfied with the result. 

There are still inconsistencies with the indentation of the code samples (I pointed them out before, but it looks like at least some of them were skipped). Unlike Python, inconsistent indentation does not affect code execution. Although I would still encourage you to format the code according to Go conventions, I certainly wouldn't insist on holding up all the other great changes to the Dev Guide just to do that. 

My only other nit is in the text just below the "Introduction to the Temporal Go SDK" heading, which says "The open source Temporal Go SDK" twice in just three sentences. I think a lot of developers would see those two mentions of "open source Temporal Go SDK" and conclude that there must also be a closed source Temporal Go SDK. I would recommend omitting "open source" here. 

If you really want to mention open source, I think it would be more effective on the "About the SDKs" page, which could say, "All Temporal SDKs, just like the Temporal software you can use to run your own cluster, are 100% open source."

flossypurse commented:
@tomwheeler I think I addressed both of your points:
I found 1 code sample that had bad indents and moved the "open source" adjective to the about sdk intro and removed it from the other sections.
